packages/form-layout/src/vaadin-form-item.js

```diff
@@ -182,14 +241,18 @@ class FormItem extends ThemableMixin(PolymerElement) {
 
   /** @private */
   __onContentSlotChange() {
+    if (this.__labelNode) {
+      this.__linkLabelToField(this.__labelNode);
+    }
+
     if (this.__contentField) {
       // Discard the old field
       this.__updateRequiredState(false);
       this.__contentFieldObserver.disconnect();
       delete this.__contentField;
     }
 
-    const contentFields = this.$.contentSlot.assignedElements().filter((node) => !!this.__getValidateFunction(node));
+    const contentFields = this.__fieldNodes.filter((node) => !!this.__getValidateFunction(node));
```

Comment:
Since `__fieldNodes` uses querySelectorAll, contentFields would end up having multiple nodes in case you use a `<vaadin-custom-field>` as the form item:
```html
<vaadin-form-item>
  <label slot="label">Range</label>
  <vaadin-custom-field required>
    <vaadin-date-picker placeholder="Start date"></vaadin-date-picker>
    &ndash;
    <vaadin-date-picker placeholder="End date"></vaadin-date-picker>
  </vaadin-custom-field>
</vaadin-form-item>
```

so the next line would assume the item has more than one field, and the required indicator would no longer work.

Reply:
Ah, @web-padawan already covered this in https://github.com/vaadin/web-components/pull/3088/files#r754861570